packages/kit/src/runtime/server/page/load_node.js

```diff
@@ -300,12 +335,148 @@ export async function load_node({
+ * @param {import('types/hooks').RequestEvent} event
+ * @param {boolean} prerender
+ * @returns {Promise<import('types/endpoint').ShadowData>}
+ */
+async function load_shadow_data(route, event, prerender) {
+	if (!route.shadow) return {};
+
+	try {
+		const mod = await route.shadow();
+
+		if (prerender && (mod.post || mod.put || mod.del || mod.patch)) {
+			throw new Error('Cannot prerender pages that have shadow endpoints with mutative methods');
+		}
+
+		const method = event.request.method.toLowerCase().replace('delete', 'del');
+		const handler = mod[method];
+
+		if (!handler) {
+			// TODO figure out what to do here. is 405 Method Not Allowed
+			// appropriate, or do we fall through?
+			return {
+				status: 405,
+				error: new Error(`${method} method not allowed`)
+			};
+		}
+
+		/** @type {import('types/endpoint').ShadowData} */
+		const data = {
+			status: 200,
+			cookies: [],
+			body: {}
+		};
+
+		if (method !== 'get') {
+			const result = await handler(event);
+
+			if (result.fallthrough) return result;
+
+			const { status = 200, headers = {}, body = {} } = result;
+
+			validate_shadow_output(headers, body);
+
+			if (headers['set-cookie']) {
+				/** @type {string[]} */ (data.cookies).push(...headers['set-cookie']);
+			}
+
+			// Redirects are respected...
+			if (status >= 300 && status < 400) {
+				return {
+					status,
+					redirect: /** @type {string} */ (
+						headers instanceof Headers ? headers.get('location') : headers.location
+					)
+				};
+			}
+
+			// ...but 4xx and 5xx status codes _don't_ result in the error page
+			// rendering for non-GET requests — instead, we allow the page
+			// to render with any validation errors etc that were returned
+			data.status = status;
+			data.body = body;
+		}
+
+		if (mod.get) {
+			const result = await mod.get.call(null, event);
+
+			if (result.fallthrough) return result;
+
+			const { status = 200, headers = {}, body = {} } = result;
+
+			validate_shadow_output(headers, body);
+
+			if (headers['set-cookie']) {
+				/** @type {string[]} */ (data.cookies).push(...headers['set-cookie']);
+			}
+
+			if (status >= 400) {
+				return {
+					status,
+					error: new Error('Failed to load data')
+				};
+			}
+
+			if (status >= 300) {
+				return {
+					status,
+					redirect: /** @type {string} */ (
+						headers instanceof Headers ? headers.get('location') : headers.location
+					)
+				};
+			}
+
+			data.body = { ...body, ...data.body };
+		}
+
+		return data;
+	} catch (e) {
+		return {
+			status: 500,
+			error: coalesce_to_error(e)
+		};
+	}
+}
+
+/**
+ * @param {Headers | Partial<import('types/helper').ResponseHeaders>} headers
+ * @param {import('types/helper').JSONValue} body
+ */
+function validate_shadow_output(headers, body) {
+	if (headers instanceof Headers && headers.has('set-cookie')) {
+		throw new Error(
+			'Shadow endpoint request handler cannot use Headers interface with Set-Cookie headers'
```

Comment:
Is this a security issue? a technical limitation? I don't see why this would be disallowed.

Reply:
Because you can put multiple `set-cookie` headers _into_ `Headers`, but you can't get them out, because Chrome engineers are actively hostile to server-side use cases for `Headers` and have blocked spec changes that would make it possible